File: macos/AgentVideoParse/Sources/FrameSampler.swift
import Foundation

enum FrameSampler {
    static func sampleTimes(
        durationSeconds: Double,
        fps: Double = AVPConstants.defaultSampleFPS,
        maxFrames: Int = AVPConstants.maxFrames
    ) -> [Double] {
        guard durationSeconds >= 0, fps > 0, maxFrames > 0 else { return [] }
        if durationSeconds == 0 { return [0] }

        let interval = 1.0 / fps
        var times: [Double] = []
        var t = 0.0
        while t < durationSeconds {
            times.append((t * 1_000_000).rounded() / 1_000_000)
            t += interval
            if times.count > maxFrames * 4 { break }
        }
        times = times.filter { $0 >= 0 && $0 <= durationSeconds }
        if times.isEmpty { times = [0] }
        if times.count > maxFrames {
            times = uniformThin(times, maxFrames: maxFrames)
        }
        var cleaned: [Double] = []
        for x in times {
            if cleaned.isEmpty || x > cleaned.last! {
                cleaned.append(x)
            }
        }
        return Array(cleaned.prefix(maxFrames))
    }

    private static func uniformThin(_ times: [Double], maxFrames: Int) -> [Double] {
        if maxFrames <= 1 { return [times[0]] }
        let n = times.count
        if n <= maxFrames { return times }
        var result: [Double] = []
        for i in 0..<maxFrames {
            let idx = Int((Double(i) * Double(n - 1) / Double(maxFrames - 1)).rounded())
            result.append(times[idx])
        }
        var out: [Double] = []
        for x in result {
            if out.isEmpty || x > out.last! { out.append(x) }
        }
        return out
    }
}

File: ui/windows/FrameSampler.cs
using System;
using System.Collections.Generic;

namespace AgentVideoParse
{
    internal static class FrameSampler
    {
        public static List<double> SampleTimes(
            double durationSeconds,
            double fps = AvpConstants.DefaultSampleFps,
            int maxFrames = AvpConstants.MaxFrames)
        {
            var times = new List<double>();
            if (durationSeconds < 0 || fps <= 0 || maxFrames <= 0)
                return times;
            if (durationSeconds == 0.0)
            {
                times.Add(0.0);
                return times;
            }

            double interval = 1.0 / fps;
            double t = 0.0;
            while (t < durationSeconds)
            {
                times.Add(Math.Round(t, 6));
                t += interval;
                if (times.Count > maxFrames * 4)
                    break;
            }

            if (times.Count > 0)
            {
                double end = Math.Max(0.0, durationSeconds - 0.001);
                if (end - times[times.Count - 1] >= interval * 0.45)
                    times.Add(Math.Round(end, 6));
            }

            times.RemoveAll(x => x < 0.0 || x > durationSeconds);
            if (times.Count == 0)
                times.Add(0.0);

            if (times.Count > maxFrames)
                times = UniformThin(times, maxFrames);

            var cleaned = new List<double>();
            foreach (var x in times)
            {
                if (cleaned.Count == 0 || x > cleaned[cleaned.Count - 1])
                    cleaned.Add(x);
            }
            if (cleaned.Count > maxFrames)
                cleaned = cleaned.GetRange(0, maxFrames);
            return cleaned;
        }

        private static List<double> UniformThin(List<double> times, int maxFrames)
        {
            if (maxFrames <= 1)
                return new List<double> { times[0] };
            int n = times.Count;
            if (n <= maxFrames)
                return times;
            var result = new List<double>();
            for (int i = 0; i < maxFrames; i++)
            {
                int idx = (int)Math.Round(i * (n - 1) / (double)(maxFrames - 1));
                result.Add(times[idx]);
            }
            var outList = new List<double>();
            foreach (var x in result)
            {
                if (outList.Count == 0 || x > outList[outList.Count - 1])
                    outList.Add(x);
            }
            return outList;
        }
    }
}
